provider/gateway/client.go

```diff
@@ -69,33 +119,92 @@ func (err ClientResponseError) ClientError() string {
 	return fmt.Sprintf("Remote Server returned %d\n%s", err.Status, err.Message)
 }
 
-func (c *client) Status(ctx context.Context, host string) (*provider.Status, error) {
-	uri, err := makeURI(host, statusPath())
+func (c *client) verifyPeerCertificate(certificates [][]byte, _ [][]*x509.Certificate) error {
+	if len(certificates) != 1 {
+		return errors.Errorf("tls: invalid certificate chain")
+	}
+
+	cert, err := x509.ParseCertificate(certificates[0])
 	if err != nil {
-		return nil, err
+		return errors.Wrap(err, "failed to parse certificate")
 	}
-	var obj provider.Status
 
-	if err := c.getStatus(ctx, uri, &obj); err != nil {
-		return nil, err
+	// validation
+	// 1. CommonName in issuer and Subject must be the same
+	if cert.Subject.CommonName != cert.Issuer.CommonName {
+		return errors.Wrap(err, "invalid certificate")
 	}
 
-	return &obj, nil
+	var prov sdk.Address
+	if prov, err = sdk.AccAddressFromBech32(cert.Subject.CommonName); err != nil {
+		return errors.Wrap(err, "invalid certificate")
+	}
+
+	if !c.addr.Equals(prov) {
+		return errors.Errorf("tls: hijacked certificate")
+	}
+
+	// 2. serial number must be in
+	if cert.SerialNumber == nil {
+		return errors.Wrap(err, "invalid certificate")
+	}
+
+	// 3. look up certificate on chain. it must not be revoked
+	var resp *ctypes.QueryCertificatesResponse
+	resp, err = c.qclient.Certificates(
+		context.Background(),
+		&ctypes.QueryCertificatesRequest{
+			Filter: ctypes.CertificateFilter{
+				Owner:  prov.String(),
+				Serial: cert.SerialNumber.String(),
+				State:  "valid",
+			},
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	certPool := x509.NewCertPool()
+
+	if !certPool.AppendCertsFromPEM(resp.Certificates[0].Cert) {
+		return errors.Wrap(err, "invalid certificate")
+	}
+
+	opts := x509.VerifyOptions{
+		Roots:                     certPool,
+		CurrentTime:               time.Now(),
+		KeyUsages:                 []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		MaxConstraintComparisions: 0,
+	}
+
+	if _, err = cert.Verify(opts); err != nil {
+		return errors.Wrap(err, "invalid certificate")
+	}
+
+	return nil
 }
 
-func createClientResponseErrorIfNotOK(resp *http.Response, responseBuf *bytes.Buffer) error {
-	if resp.StatusCode == http.StatusOK {
-		return nil
+func (c *client) verifyConnection(_ tls.ConnectionState) error {
```

Comment:
I think you can just set `verifyConnection: nil` in the TLS configuration if this method does not need to do anything